src/parser.rs

```diff
@@ -0,0 +1,152 @@
+#![allow(dead_code)]
+use crate::lexer::Token;
+use std::fmt;
+use std::iter::Iterator;
+
+#[derive(PartialEq, Debug)]
+pub enum Expression {
+    Identifier(String),
+    Number(i64),
+    ProcudureCall(Box<Expression>, Vec<Box<Expression>>),
+}
+
+#[derive(Debug)]
+pub struct SyntaxError {
+    error: String,
+}
+
+impl fmt::Display for SyntaxError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Syntax error: {}", self.error)
+    }
+}
+
+macro_rules! syntax_error {
+    ($($arg:tt)*) => (
+        return Err(SyntaxError { error: format!($($arg)*) });
+    )
+}
+
+pub struct Parser<TokenIter: Iterator<Item = Token> + Clone> {
+    current: Option<Token>,
+    lexer: TokenIter,
+}
+
+impl<TokenIter: Iterator<Item = Token> + Clone> Parser<TokenIter> {
+    pub fn new(mut lexer: TokenIter) -> Parser<TokenIter> {
+        Self {
+            current: lexer.next(),
+            lexer: lexer,
+        }
+    }
+
+    pub fn parse(&mut self) -> Result<Option<Box<Expression>>, SyntaxError> {
+        match self.current.clone() {
+            Some(token) => match token {
+                Token::Number(a) => self.generate(Box::new(Expression::Number(a))),
+                Token::Identifier(a) => self.generate(Box::new(Expression::Identifier(a))),
+                Token::LeftParen => self.procedure_call(),
+                _ => Ok(None),
+            },
+            None => Ok(None),
+        }
+    }
+
+    fn procedure_call(&mut self) -> Result<Option<Box<Expression>>, SyntaxError> {
+        self.advance();
+        let operator = self.parse()?.unwrap();
+        let mut params: Vec<Box<Expression>> = vec![];
+        loop {
+            match &self.current {
+                Some(Token::RightParen) => {
+                    return self.generate(Box::new(Expression::ProcudureCall(operator, params)));
+                }
+                None => syntax_error!("Un matched parentheses!"),
+                _ => params.push(self.parse()?.unwrap()),
+            }
+        }
+    }
+
+    fn advance(&mut self) {
+        self.current = self.lexer.next();
+    }
+
+    fn generate(&mut self, ast: Box<Expression>) -> Result<Option<Box<Expression>>, SyntaxError> {
+        self.advance();
+        Ok(Some(ast))
+    }
+}
+
+#[test]
+fn empty() {
+    let empty_tokens: Vec<Token> = Vec::new();
+    let mut parser = Parser::new(empty_tokens.into_iter());
+    let ast = parser.parse().unwrap();
+    assert_eq!(ast, None);
+}
+
+#[test]
+fn number() {
+    let empty_tokens: Vec<Token> = vec![Token::Number(1)];
```

Comment:
It is not `empty_tokens`. They may be named simply `tokens` for all tests.

Reply:
Yes, will update later.

Reply:
Updated